### DISCUSSION TITLE
Inference c++ jetson xavier AGX

### DISCUSSION THREAD

**Discussion by christophezeinaty** [larq/compute-engine#782] [General]
Hello I trained my model using larq and saved the model using this 

` with open("bnn.tflite", 'wb') as flatbuffer_file:
        flatbuffer_bytes = lce.convert_keras_model(model_b)
        flatbuffer_file.write(flatbuffer_bytes)`

I worte a c++ code to do the inference using the LCE engine and compiled the code directly on the jetson xavier which is arm64 based processor. I am not sure if my inference uses matrix operations or is it really based only on bitwise operations. Whats bothering me is that this model has 76 000 params and do a 60 ms inference time per image, if I compare this model to a smaller one with 12 000 params and quantized weights int 8 the BNN is slower. The int8 models takes only 5ms per image is it normal ? 

My hypothesis was even if I have a model that is a little bit larger but is binarized it should be faster (due to removing the mat mult operations) , am I wrong ?

**Comment by Tombana** [comment]
Hi @christophezeinaty, could you open the `bnn.tflite` file in https://netron.app/ and show a screenshot of it? Then we can check if it was converted correctly. (And maybe also a screenshot of the int8 tflite file, for comparison)
Since the Jetson is designed for AI, it is possible that the int8 model is accelerated by either the GPU or by special int8 SIMD instructions on the CPU, which might not exist for binary operations. So although the binary model might be faster in theory, it could be slower on hardware that was specifically designed for int8 instead of binary.

**Reply by christophezeinaty** [reply]
here is a link to see the arch [bnn_arch](https://docs.google.com/document/d/1sGgwU1hbxoL6hOARVAvDdXP_VHpt0cI3gERJe66afMI/edit) of the bnn model, the int8 I dont have it, I only have the results for the same data set but I know in fact that its a smaller model based on depthwise conv.
So how would I test that there is a potential in BNN compared to that model not only theoretically ?

**Reply by Tombana** [reply]
The converted model screenshots look ok. In general it is difficult to compare different model architectures like that, for example a model with depthwise layers versus regular layers. Especially the number of parameters is _not_ a good proxy for inference time. Perhaps you can take the BNN architecture but remove the binary quantizers and use it as an int8 model instead, so that everything is the same except for int8 vs binary. Then you can more directly compare the inference times.

**Reply by christophezeinaty** [reply]
An I checked rapidly it doesn't seem that the jetson has something special for the int8

**Reply by christophezeinaty** [reply]
but my goal is to show the advantage of using binary vs int8 in inference more generally to show that removing matrix mult can give a benefits

**Reply by Tombana** [reply]
Note that there are still matrix multiplications: the only difference is that multiplying the entries of the matrix (which are now single bits instead of int8 values) don't require int8 or float multiply operations but only binary XOR and popcount operations.

**Reply by christophezeinaty** [reply]
that's my c++ code should I do something special to enable the bitwise ops or this is correct

`InferenceResult performInference(const char* model_path, const char* image_path) {
    InferenceResult result;

    // Load the model
    std::unique_ptr<tflite::FlatBufferModel> model = tflite::FlatBufferModel::BuildFromFile(model_path);
    if (!model) {
        std::cerr << "Failed to load model." << std::endl;
        return result;
    }
    
    // create a builtin OpResolver
    tflite::ops::builtin::BuiltinOpResolver resolver;
    // register LCE custom ops
    compute_engine::tflite::RegisterLCECustomOps(&resolver);
    // Build the interpreter
    tflite::InterpreterBuilder builder(*model, resolver);
    std::unique_ptr<tflite::Interpreter> interpreter;

    if (builder(&interpreter) != kTfLiteOk) {
        std::cerr << "Failed to build interpreter." << std::endl;
        return result;
    }

    // Allocate tensors for the model
    if (interpreter->AllocateTensors() != kTfLiteOk) {
        std::cerr << "Failed to allocate tensors." << std::endl;
        return result;
    }

    int width, height, channels;
    unsigned char* image_data = stbi_load(image_path, &width, &height, &channels, 0);
    if (!image_data) {
        std::cerr << "Failed to load image." << std::endl;
        return result;
    }


    float* normalized_image_data = new float[width * height * channels];
    for (int i = 0; i < width * height * channels; ++i) {
        normalized_image_data[i] = (static_cast<float>(image_data[i]) / 255.0f - 0.5) * 2;
    }

    const float* input_tensor = interpreter->typed_input_tensor<float>(0);
    int input_width = interpreter->input_tensor(0)->dims->data[2];
    int input_height = interpreter->input_tensor(0)->dims->data[1];
    int input_channels = interpreter->input_tensor(0)->dims->data[3];

    memcpy(interpreter->typed_input_tensor<float>(0), normalized_image_data, input_width * input_height * input_channels * sizeof(float));

    // Run inference
    auto start_time = std::chrono::high_resolution_clock::now();
    if (interpreter->Invoke() != kTfLiteOk) {
        std::cerr << "Failed to invoke interpreter." << std::endl;
        return result;
    }
    auto end_time = std::chrono::high_resolution_clock::now();
    result.inference_time = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time);
}`

**Reply by Tombana** [reply]
The code looks correct. I recommend taking the same model with and without the binary quantizers to be able to directly compare performance.

**Reply by christophezeinaty** [reply]
Not really sure I understood, if I remove binary quant it will be just comparing int8 to a normal model based on CNN, thats not what I want to do, can I still save my bnn model weights with int8?

**Reply by Tombana** [reply]
I meant it as a test to see if binary inference is working as expected. The binary model should be a lot faster than the non-binary model. If it's not faster then we can conclude that something is broken somewhere.
Directly comparing the binary model with a different int8 model that has a different architecture does not always make sense, there could be a lot of reasons why it is slower or faster. Perhaps the int8 model has much fewer ops for example.

**Reply by christophezeinaty** [reply]
ah ok I see I will test that and will keep you updated, thank you for your time.

**Reply by christophezeinaty** [reply]
ok I tested with same arch but without the BNN quantization and clearly there is an issue, since the new model (without BNN) takes 10ms per image compared to the BNN one that takes 65ms per image.

**Reply by Tombana** [reply]
In your code I see that you run `Invoke()` once. Can you try to run it multiple times? The binary convolution kernel does extra setup work in the first iteration, so the first iteration is slower than the others.

You can also try to add a print statement here to see if it is actually running the correct code:

https://github.com/larq/compute-engine/blob/d8dab64743d8564f36bf0f145a70d4bbcc88453b/larq_compute_engine/tflite/kernels/bconv2d.cc#L417-L421

**Reply by christophezeinaty** [reply]
ok so I added another invoke there is a small improvement instead of 65ms its 45ms but still slower than the non BNN  model. I added the print as well seems the code is not calling this function

**Reply by Tombana** [reply]
Perhaps `TFLITE_WITH_RUY` is not enabled when larq-compute-engine is compiled, then it might fall back to a slow reference implementation.
Can you try to add `set(TFLITE_ENABLE_RUY ON CACHE BOOL "Enable RUY")` to `CMakeLists.txt` before including the larq-compute-engine cmake file?

**Reply by christophezeinaty** [reply]
I added it like this nothing changed:

`set(TFLITE_ENABLE_RUY ON CACHE BOOL "Enable RUY")
set(COMPUTE_ENGINE_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/compute-engine CACHE PATH
  "Directory that contains the compute-engine project"
)
if(NOT COMPUTE_ENGINE_SOURCE_DIR)
  get_filename_component(COMPUTE_ENGINE_SOURCE_DIR
    "${CMAKE_CURRENT_LIST_DIR}/../../../../"
    ABSOLUTE
  )
endif()

add_subdirectory(
  "${COMPUTE_ENGINE_SOURCE_DIR}/"
  "${CMAKE_CURRENT_BINARY_DIR}/lq_build"
  EXCLUDE_FROM_ALL
)

set(CMAKE_CXX_STANDARD 17)
add_executable(dicetrack
  main.cpp
)
target_include_directories(dicetrack PRIVATE 
${CMAKE_CURRENT_SOURCE_DIR}/includes
${CMAKE_CURRENT_SOURCE_DIR}/compute-engine)

target_link_libraries(dicetrack
larq-compute-engine 
)
`

**Reply by Tombana** [reply]
That's unfortunate. Perhaps `CMakeCache.txt` needs to be deleted and recreated, it might have remembered some old settings.
If that doesn't work, you can also try `add_compile_definitions(TFLITE_WITH_RUY)`.

**Reply by christophezeinaty** [reply]
I just cleared the CMakeCache.txt and now its working ! the BNN mode is doing an inference of 14ms per image(using 1 Invoke,if I use 2 the second will be 5ms) but is it normal that its approx the same execution time as the non BNN model ?

**Reply by Tombana** [reply]
Good to hear that it works!
So if I understand correctly, its now 5 ms for the BNN (second Invoke) and 10 ms for the non-BNN model with the same architecture?
I agree that this is not a great speedup (only factor 2X). It could be that the NVIDIA Carmel CPU is very optimized for float and int8 SIMD instructions, but not so much for BNN-related instructions like xor and popcount. Unfortunately there's not much we can do here.
One last thing you can try is to set `use_indirect_bgemm` to `true`:
```cpp
const bool use_reference_bconv = false;
const bool use_indirect_bgemm = true;
compute_engine::tflite::RegisterLCECustomOps(&resolver, use_reference_bconv, use_indirect_bgemm);
```

**Reply by christophezeinaty** [reply]
actually there is no speedup at all its exactly the same for both. (first invoke 10ms, 5ms for second)

**Reply by christophezeinaty** [reply]
and adding this 

`const bool use_reference_bconv = false;
const bool use_indirect_bgemm = true;
compute_engine::tflite::RegisterLCECustomOps(&resolver, use_reference_bconv, use_indirect_bgemm);`

Doubled the inference time

**Reply by Tombana** [reply]
In that case I'm afraid I don't have any other ideas. Perhaps this hardware platform is simply not well-suited for BNNs.

**Reply by christophezeinaty** [reply]
that's  really weird since the Jetson Xavier AGX is an arm64 which is compatible with the larq documentation. Unless the Jetson is really optimized with the fp and int8 which is unfortunate that means using hardware we can have faster inference.

Anyway thank you a lot for your help (it was like real time support which is rare)

**Reply by christophezeinaty** [reply]
@Tombana small update I tried same tests on RPI4 and I have no improvement between bnn_model and same architecture not bnn. In fact on the RPI the one without BNN layers is faster

**Reply by Tombana** [reply]
Thanks for letting us know. Can you try the latest `lce_benchmark_model_aarch64` from https://github.com/larq/compute-engine/releases ? It should work on a 64-bit arm64 machine. You can get per-layer timings from that as well, then maybe we can debug the issue more easily. It's used as:
```
./lce_benchmark_model --graph=<path_to_tflite> --enable_op_profiling=true
```
Optionally you can also use `--use_xnnpack=true`.

**Reply by christophezeinaty** [reply]
I have thhis missing GLIBC could that be something affecting my results ? 

`/lce_benchmark_model_aarch64 --graph=../models/bnn.tflite --enable_op_profiling=true'

'./lce_benchmark_model_aarch64: /lib/aarch64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by ./lce_benchmark_model_aarch64)'

'./lce_benchmark_model_aarch64: /lib/aarch64-linux-gnu/libc.so.6: version `GLIBC_2.32' not found (required by ./lce_benchmark_model_aarch64)'

'./lce_benchmark_model_aarch64: /lib/aarch64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found (required by ./lce_benchmark_model_aarch64)'

.'/lce_benchmark_model_aarch64: /lib/aarch64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.29' not found (required by ./lce_benchmark_model_aarch6)`

**Reply by christophezeinaty** [reply]
this is the version of GLIBC that I have ldd (Debian GLIBC 2.31-13+rpt2+rpi1+deb11u5) 2.3

**Reply by Tombana** [reply]
Oh right I forgot about that, it requires a newer OS. In that case it might be easier to compile `lce_benchmark_model` yourself. You already have the correct cmake setup, all you have to do is build `lce_benchmark_model` in cmake. (For example `make lce_benchmark_model` if you use the `make` backend.)

**Reply by christophezeinaty** [reply]
`Here are the results for the BNN model `
INFO: STARTING!
WARN: Duplicate flags: num_threads
INFO: Log parameter values verbosely: [0]
INFO: Graph: [../../../models/bnn.tflite]
INFO: Enable op profiling: [1]
INFO: Loaded model ../../../models/bnn.tflite
INFO: The input model file size (MB): 0.04172
INFO: Initialized session in 7.126ms.
INFO: Running benchmark for at least 1 iterations and at least 0.5 seconds but terminate if exceeding 150 seconds.
INFO: count=3 first=266695 curr=219260 min=218874 max=266695 avg=234943 std=22452

INFO: Running benchmark for at least 50 iterations and at least 1 seconds but terminate if exceeding 150 seconds.
INFO: count=50 first=219389 curr=219173 min=218465 max=219509 avg=219151 std=182

INFO: Inference timings in us: Init: 7126, First inference: 266695, Warmup (avg): 234943, Inference (avg): 219151
INFO: Note: as the benchmark tool itself affects memory footprint, the following is only APPROXIMATE to the actual memory footprint of the model at runtime. Take the information at your discretion.
INFO: Memory footprint delta from the start of the tool (MB): init=2.75391 overall=4.16797
INFO: Profiling Info for Benchmark Initialization:
============================== Run Order ==============================
	                             [node type]	  [first]	 [avg ms]	     [%]	  [cdf%]	  [mem KB]	[times called]	[Name]
	                 ModifyGraphWithDelegate	    1.724	    1.724	 84.098%	 84.098%	     0.000	        1	ModifyGraphWithDelegate/0
	                         AllocateTensors	    0.326	    0.326	 15.902%	100.000%	     0.000	        1	AllocateTensors/0

============================== Top by Computation Time ==============================
	                             [node type]	  [first]	 [avg ms]	     [%]	  [cdf%]	  [mem KB]	[times called]	[Name]
	                 ModifyGraphWithDelegate	    1.724	    1.724	 84.098%	 84.098%	     0.000	        1	ModifyGraphWithDelegate/0
	                         AllocateTensors	    0.326	    0.326	 15.902%	100.000%	     0.000	        1	AllocateTensors/0

Number of nodes executed: 2
============================== Summary by node type ==============================
	                             [Node type]	  [count]	  [avg ms]	    [avg %]	    [cdf %]	  [mem KB]	[times called]
	                 ModifyGraphWithDelegate	        1	     1.724	    84.098%	    84.098%	     0.000	        1
	                         AllocateTensors	        1	     0.326	    15.902%	   100.000%	     0.000	        1

Timings (microseconds): count=1 curr=2050
Memory (bytes): count=0
2 nodes observed



INFO: Operator-wise Profiling Info for Regular Benchmark Runs:
============================== Run Order ==============================
	                             [node type]	  [first]	 [avg ms]	     [%]	  [cdf%]	  [mem KB]	[times called]	[Name]
	           Convolution (NHWC, F32) IGEMM	    0.537	    0.536	  0.245%	  0.245%	     0.000	        1	Delegate/Convolution (NHWC, F32) IGEMM:0
	                             LceQuantize	    2.228	    2.215	  1.011%	  1.256%	     0.000	        1	[sequential/quant_conv2d_1/ste_sign_2/Sign_1;sequen1]:1
	                              LceBconv2d	   83.141	   83.054	 37.911%	 39.167%	     0.000	        1	[sequential/quant_conv2d_1/QuantConv2D;sequential/q2]:2
	                 Max Pooling (NHWC, F32)	    0.078	    0.038	  0.052%	 39.219%	     0.000	        3	Delegate/Max Pooling (NHWC, F32):0
	                      Multiply (ND, F32)	    0.023	    0.012	  0.017%	 39.235%	     0.000	        3	Delegate/Multiply (ND, F32):1
	                           Add (ND, F32)	    0.023	    0.012	  0.016%	 39.252%	     0.000	        3	Delegate/Add (ND, F32):2
	                             LceQuantize	    0.556	    0.554	  0.253%	 39.504%	     0.000	        1	[sequential/quant_conv2d_2/ste_sign_4/Sign_1;sequen]:6
	                              LceBconv2d	   40.738	   40.757	 18.604%	 58.108%	     0.000	        1	[sequential/batch_normalization_2/FusedBatchNormV3;1]:7
	                             LceQuantize	    0.015	    0.015	  0.007%	 58.115%	     0.000	        1	[sequential/quant_conv2d_3/ste_sign_6/Sign_1;sequen]:8
	                              LceBconv2d	   40.824	   40.759	 18.605%	 76.720%	     0.000	        1	[sequential/quant_conv2d_3/QuantConv2D;sequential/q1]:9
	                       Softmax (NC, F32)	    0.002	    0.001	  0.001%	 76.721%	     0.000	        1	Delegate/Softmax (NC, F32):1
	                           LceDequantize	    0.002	    0.002	  0.001%	 76.721%	     0.000	        1	[sequential/quant_dense_1/ste_sign_14/Sign_1;sequen1]:25
	                             LceQuantize	    0.002	    0.001	  0.001%	 76.722%	     0.000	        1	[sequential/quant_dense_1/ste_sign_14/Sign_1;sequen]:24
	          Fully Connected (NC, F32) GEMM	    0.008	    0.003	  0.003%	 76.725%	     0.000	        2	Delegate/Fully Connected (NC, F32) GEMM:0
	                           LceDequantize	    0.003	    0.002	  0.001%	 76.726%	     0.000	        1	[sequential/quant_dense/ste_sign_12/Sign_1;sequenti1]:22
	                             LceQuantize	    0.007	    0.005	  0.002%	 76.728%	     0.000	        1	[sequential/quant_conv2d_4/ste_sign_8/Sign_1;sequen]:13
	                              LceBconv2d	   18.376	   18.335	  8.369%	 85.098%	     0.000	        1	[sequential/batch_normalization_4/FusedBatchNormV3;1]:14
	                             LceQuantize	    0.007	    0.008	  0.003%	 85.101%	     0.000	        1	[sequential/quant_conv2d_5/ste_sign_10/Sign_1;seque]:15
	                              LceBconv2d	   32.650	   32.636	 14.897%	 99.998%	     0.000	        1	[sequential/quant_conv2d_5/QuantConv2D;sequential/q1]:16
	       Global Average Pooling (NWC, F32)	    0.003	    0.002	  0.001%	 99.999%	     0.000	        1	Delegate/Global Average Pooling (NWC, F32):3
	                             LceQuantize	    0.002	    0.002	  0.001%	100.000%	     0.000	        1	[sequential/quant_dense/ste_sign_12/Sign_1;sequenti]:21

============================== Top by Computation Time ==============================
	                             [node type]	  [first]	 [avg ms]	     [%]	  [cdf%]	  [mem KB]	[times called]	[Name]
	                              LceBconv2d	   83.141	   83.054	 37.911%	 37.911%	     0.000	        1	[sequential/quant_conv2d_1/QuantConv2D;sequential/q2]:2
	                              LceBconv2d	   40.824	   40.759	 18.605%	 56.516%	     0.000	        1	[sequential/quant_conv2d_3/QuantConv2D;sequential/q1]:9
	                              LceBconv2d	   40.738	   40.757	 18.604%	 75.120%	     0.000	        1	[sequential/batch_normalization_2/FusedBatchNormV3;1]:7
	                              LceBconv2d	   32.650	   32.636	 14.897%	 90.017%	     0.000	        1	[sequential/quant_conv2d_5/QuantConv2D;sequential/q1]:16
	                              LceBconv2d	   18.376	   18.335	  8.369%	 98.386%	     0.000	        1	[sequential/batch_normalization_4/FusedBatchNormV3;1]:14
	                             LceQuantize	    2.228	    2.215	  1.011%	 99.397%	     0.000	        1	[sequential/quant_conv2d_1/ste_sign_2/Sign_1;sequen1]:1
	                             LceQuantize	    0.556	    0.554	  0.253%	 99.650%	     0.000	        1	[sequential/quant_conv2d_2/ste_sign_4/Sign_1;sequen]:6
	           Convolution (NHWC, F32) IGEMM	    0.537	    0.536	  0.245%	 99.895%	     0.000	        1	Delegate/Convolution (NHWC, F32) IGEMM:0
	                 Max Pooling (NHWC, F32)	    0.078	    0.038	  0.052%	 99.947%	     0.000	        3	Delegate/Max Pooling (NHWC, F32):0
	                             LceQuantize	    0.015	    0.015	  0.007%	 99.954%	     0.000	        1	[sequential/quant_conv2d_3/ste_sign_6/Sign_1;sequen]:8

Number of nodes executed: 21
============================== Summary by node type ==============================
	                             [Node type]	  [count]	  [avg ms]	    [avg %]	    [cdf %]	  [mem KB]	[times called]
	                              LceBconv2d	        5	   215.541	    98.390%	    98.390%	     0.000	        5
	                             LceQuantize	        7	     2.796	     1.276%	    99.666%	     0.000	        7
	           Convolution (NHWC, F32) IGEMM	        1	     0.535	     0.244%	    99.910%	     0.000	        1
	                 Max Pooling (NHWC, F32)	        1	     0.114	     0.052%	    99.962%	     0.000	        3
	                      Multiply (ND, F32)	        1	     0.036	     0.016%	    99.979%	     0.000	        3
	                           Add (ND, F32)	        1	     0.035	     0.016%	    99.995%	     0.000	        3
	          Fully Connected (NC, F32) GEMM	        1	     0.006	     0.003%	    99.997%	     0.000	        2
	                           LceDequantize	        2	     0.003	     0.001%	    99.999%	     0.000	        2
	       Global Average Pooling (NWC, F32)	        1	     0.002	     0.001%	   100.000%	     0.000	        1
	                       Softmax (NC, F32)	        1	     0.001	     0.000%	   100.000%	     0.000	        1

Timings (microseconds): count=50 first=219294 curr=219106 min=218393 max=219432 avg=219077 std=179
Memory (bytes): count=0
21 nodes observed

**Reply by christophezeinaty** [reply]
and these are for the no_bnn (which is crazy it looks faster) note these test are done on the RPI4 not the jetson:

INFO: STARTING!
WARN: Duplicate flags: num_threads
INFO: Log parameter values verbosely: [0]
INFO: Graph: [../../models/no_bnn.tflite]
INFO: Enable op profiling: [1]
INFO: Loaded model ../../models/no_bnn.tflite
INFO: The input model file size (MB): 0.315872
INFO: Initialized session in 8.719ms.
INFO: Running benchmark for at least 1 iterations and at least 0.5 seconds but terminate if exceeding 150 seconds.
INFO: count=51 first=29295 curr=9400 min=9379 max=29295 avg=9803.45 std=2758

INFO: Running benchmark for at least 50 iterations and at least 1 seconds but terminate if exceeding 150 seconds.
INFO: count=104 first=9504 curr=9453 min=9430 max=11017 avg=9473.11 std=153

INFO: Inference timings in us: Init: 8719, First inference: 29295, Warmup (avg): 9803.45, Inference (avg): 9473.11
INFO: Note: as the benchmark tool itself affects memory footprint, the following is only APPROXIMATE to the actual memory footprint of the model at runtime. Take the information at your discretion.
INFO: Memory footprint delta from the start of the tool (MB): init=3.69141 overall=5.5
INFO: Profiling Info for Benchmark Initialization:
============================== Run Order ==============================
	                             [node type]	  [first]	 [avg ms]	     [%]	  [cdf%]	  [mem KB]	[times called]	[Name]
	                 ModifyGraphWithDelegate	    3.581	    3.581	 94.836%	 94.836%	   740.000	        1	ModifyGraphWithDelegate/0
	                         AllocateTensors	    0.195	    0.195	  5.164%	100.000%	     0.000	        1	AllocateTensors/0

============================== Top by Computation Time ==============================
	                             [node type]	  [first]	 [avg ms]	     [%]	  [cdf%]	  [mem KB]	[times called]	[Name]
	                 ModifyGraphWithDelegate	    3.581	    3.581	 94.836%	 94.836%	   740.000	        1	ModifyGraphWithDelegate/0
	                         AllocateTensors	    0.195	    0.195	  5.164%	100.000%	     0.000	        1	AllocateTensors/0

Number of nodes executed: 2
============================== Summary by node type ==============================
	                             [Node type]	  [count]	  [avg ms]	    [avg %]	    [cdf %]	  [mem KB]	[times called]
	                 ModifyGraphWithDelegate	        1	     3.581	    94.836%	    94.836%	   740.000	        1
	                         AllocateTensors	        1	     0.195	     5.164%	   100.000%	     0.000	        1

Timings (microseconds): count=1 curr=3776
Memory (bytes): count=0
2 nodes observed



INFO: Operator-wise Profiling Info for Regular Benchmark Runs:
============================== Run Order ==============================
	                             [node type]	  [first]	 [avg ms]	     [%]	  [cdf%]	  [mem KB]	[times called]	[Name]
	           Convolution (NHWC, F32) IGEMM	    0.620	    0.615	  6.545%	  6.545%	     0.000	        1	Delegate/Convolution (NHWC, F32) IGEMM:0
	           Convolution (NHWC, F32) IGEMM	    2.207	    2.184	 23.261%	 29.806%	     0.000	        1	Delegate/Convolution (NHWC, F32) IGEMM:1
	                 Max Pooling (NHWC, F32)	    0.093	    0.099	  1.054%	 30.860%	     0.000	        1	Delegate/Max Pooling (NHWC, F32):2
	                      Multiply (ND, F32)	    0.026	    0.026	  0.272%	 31.133%	     0.000	        1	Delegate/Multiply (ND, F32):3
	                           Add (ND, F32)	    0.024	    0.024	  0.261%	 31.393%	     0.000	        1	Delegate/Add (ND, F32):4
	           Convolution (NHWC, F32) IGEMM	    1.135	    1.146	 12.208%	 43.601%	     0.000	        1	Delegate/Convolution (NHWC, F32) IGEMM:5
	           Convolution (NHWC, F32) IGEMM	    2.134	    2.130	 22.679%	 66.280%	     0.000	        1	Delegate/Convolution (NHWC, F32) IGEMM:6
	                 Max Pooling (NHWC, F32)	    0.032	    0.033	  0.351%	 66.631%	     0.000	        1	Delegate/Max Pooling (NHWC, F32):7
	                      Multiply (ND, F32)	    0.009	    0.009	  0.100%	 66.731%	     0.000	        1	Delegate/Multiply (ND, F32):8
	                           Add (ND, F32)	    0.010	    0.010	  0.105%	 66.836%	     0.000	        1	Delegate/Add (ND, F32):9
	           Convolution (NHWC, F32) IGEMM	    1.057	    1.065	 11.345%	 78.181%	     0.000	        1	Delegate/Convolution (NHWC, F32) IGEMM:10
	           Convolution (NHWC, F32) IGEMM	    2.009	    2.007	 21.373%	 99.554%	     0.000	        1	Delegate/Convolution (NHWC, F32) IGEMM:11
	                 Max Pooling (NHWC, F32)	    0.017	    0.017	  0.182%	 99.736%	     0.000	        1	Delegate/Max Pooling (NHWC, F32):12
	                      Multiply (ND, F32)	    0.005	    0.005	  0.054%	 99.790%	     0.000	        1	Delegate/Multiply (ND, F32):13
	                           Add (ND, F32)	    0.004	    0.004	  0.043%	 99.833%	     0.000	        1	Delegate/Add (ND, F32):14
	       Global Average Pooling (NWC, F32)	    0.003	    0.003	  0.032%	 99.865%	     0.000	        1	Delegate/Global Average Pooling (NWC, F32):15
	          Fully Connected (NC, F32) GEMM	    0.007	    0.008	  0.080%	 99.945%	     0.000	        1	Delegate/Fully Connected (NC, F32) GEMM:16
	          Fully Connected (NC, F32) GEMM	    0.004	    0.003	  0.031%	 99.977%	     0.000	        1	Delegate/Fully Connected (NC, F32) GEMM:17
	                       Softmax (NC, F32)	    0.002	    0.002	  0.023%	100.000%	     0.000	        1	Delegate/Softmax (NC, F32):18

============================== Top by Computation Time ==============================
	                             [node type]	  [first]	 [avg ms]	     [%]	  [cdf%]	  [mem KB]	[times called]	[Name]
	           Convolution (NHWC, F32) IGEMM	    2.207	    2.184	 23.261%	 23.261%	     0.000	        1	Delegate/Convolution (NHWC, F32) IGEMM:1
	           Convolution (NHWC, F32) IGEMM	    2.134	    2.130	 22.679%	 45.940%	     0.000	        1	Delegate/Convolution (NHWC, F32) IGEMM:6
	           Convolution (NHWC, F32) IGEMM	    2.009	    2.007	 21.373%	 67.313%	     0.000	        1	Delegate/Convolution (NHWC, F32) IGEMM:11
	           Convolution (NHWC, F32) IGEMM	    1.135	    1.146	 12.208%	 79.521%	     0.000	        1	Delegate/Convolution (NHWC, F32) IGEMM:5
	           Convolution (NHWC, F32) IGEMM	    1.057	    1.065	 11.345%	 90.867%	     0.000	        1	Delegate/Convolution (NHWC, F32) IGEMM:10
	           Convolution (NHWC, F32) IGEMM	    0.620	    0.615	  6.545%	 97.412%	     0.000	        1	Delegate/Convolution (NHWC, F32) IGEMM:0
	                 Max Pooling (NHWC, F32)	    0.093	    0.099	  1.054%	 98.466%	     0.000	        1	Delegate/Max Pooling (NHWC, F32):2
	                 Max Pooling (NHWC, F32)	    0.032	    0.033	  0.351%	 98.817%	     0.000	        1	Delegate/Max Pooling (NHWC, F32):7
	                      Multiply (ND, F32)	    0.026	    0.026	  0.272%	 99.089%	     0.000	        1	Delegate/Multiply (ND, F32):3
	                           Add (ND, F32)	    0.024	    0.024	  0.261%	 99.349%	     0.000	        1	Delegate/Add (ND, F32):4

Number of nodes executed: 19
============================== Summary by node type ==============================
	                             [Node type]	  [count]	  [avg ms]	    [avg %]	    [cdf %]	  [mem KB]	[times called]
	           Convolution (NHWC, F32) IGEMM	        6	     9.144	    97.474%	    97.474%	     0.000	        6
	                 Max Pooling (NHWC, F32)	        3	     0.147	     1.567%	    99.041%	     0.000	        3
	                      Multiply (ND, F32)	        3	     0.039	     0.416%	    99.456%	     0.000	        3
	                           Add (ND, F32)	        3	     0.037	     0.394%	    99.851%	     0.000	        3
	          Fully Connected (NC, F32) GEMM	        2	     0.009	     0.096%	    99.947%	     0.000	        2
	       Global Average Pooling (NWC, F32)	        1	     0.003	     0.032%	    99.979%	     0.000	        1
	                       Softmax (NC, F32)	        1	     0.002	     0.021%	   100.000%	     0.000	        1

Timings (microseconds): count=104 first=9398 curr=9371 min=9348 max=10932 avg=9390.11 std=153
Memory (bytes): count=0
19 nodes observed

**Reply by Tombana** [reply]
Could you check that this is being compiled with `-O3` optimizations (i.e. `CMAKE_BUILD_TYPE=Release`), and perhaps check with a print-statement here if this function is being called:
https://github.com/larq/compute-engine/blob/d8dab64743d8564f36bf0f145a70d4bbcc88453b/larq_compute_engine/tflite/kernels/bconv2d.cc#L417-L421

**Reply by christophezeinaty** [reply]
For my code it is indeed compiling and working we checked that before when I added this in my CMakeList 
`set(TFLITE_ENABLE_RUY ON CACHE BOOL "Enable RUY")`
But for the lce_benchmark_model I am not sure, I can confirm that when I run my code the bnn model is much faster than the one used by the benchmark but still slower than the no_bnn

**Reply by christophezeinaty** [reply]
but I build the code only with cmake .. then make

**Reply by christophezeinaty** [reply]
I realized that my model is trained with tf 2.13.0 could that be a problem since LCE uses older version of TF ?

**Reply by Tombana** [reply]
The version of tensorflow should not matter. Could you perhaps share the `.tflite` files themselves (one with and one without BNNs)?

**Reply by Tombana** [reply]
I also benchmarked them on a RPi4 and I got the following results:

The BNN was a factor 12X faster: I got `0.8 ms` for the BNN and `9.7 ms` for the non-BNN.

The logs are here: [nobnn.txt](https://github.com/larq/compute-engine/files/12251184/results_nobnn.txt) and [bnn.txt](https://github.com/larq/compute-engine/files/12251185/results_bnn.txt)

So it looks like somehow the correct BNN code is not being executed, but I'm not sure why. This `lce_benchmark` was compiled on the RPi4 which has `gcc 10.2`.

Are you sure it's not being built in debug-mode? (use `-DCMAKE_BUILD_TYPE=Release` when generating cmake)

**Reply by christophezeinaty** [reply]
Ah wow that's good news, I will try to see why I am having this problem. I built the lace compute engine without any flag so maybe by default it's debug ? And do you think that's the problem ? I will test that and will let you know thank you.

**Reply by christophezeinaty** [reply]
I tried to generate the cmake with that -DCMAKE_BUILD_TYPE=Release, it didn't work, should I do the same while building my code ? 

(but for the ./lce_benchmark_model it didnt work), its a little bit frustrating.

**Reply by christophezeinaty** [reply]
could that be a problem if I am building with Cmake. if I want to try to build with bazel how can integrate that in my CMakeList.txt (c++ code)

**Reply by Tombana** [reply]
To be sure, I tested it again on a RPi4 completely "from scratch":

- Clone `https://github.com/larq/compute-engine` and initialize submodules
  - `git clone https://github.com/larq/compute-engine`
  - `cd compute-engine`
  - `git submodule update --init --recursive --depth=1`  (EDIT: no `--` in front of `update`)
- Open `CMakeLists.txt` and add `set(TFLITE_ENABLE_RUY ON CACHE BOOL "Enable RUY")` at the start (I added it on line 8)
- Do CMake generation with `-DCMAKE_BUILD_TYPE=Release`:
  - `mkdir build`
  - `cd build`
  - `cmake ../ -DCMAKE_BUILD_TYPE=Release`
- Build all
  - `make` (or `ninja` if you use that)
- Run `./build/lce_benchmark_model --graph=bnn.tflite --enable_op_profiling=true`

I got the same timings as before.

(I did see that I looked at the wrong numbers before, the BNN was not `0.8 ms` but `2.9 ms`, so it's not a 12X speedup but only 3.3X)

**Reply by christophezeinaty** [reply]
Ok I will test that, by default in the third_party dir the tensorflow dir is empty should I colne tf in it or there is a way to do it directly ?

**Reply by Tombana** [reply]
I should've mentioned, you have to initialize submodules: `git submodule update --init --recursive`. I'll edit my post above as well. (EDIT: no `--` in front of `update`)

**Reply by christophezeinaty** [reply]
ok It worked, seems like there is something wrong with the way I build LCE, thank you a lot. And to clarify we should always look at the line were its written Inference time not the time we see in the end of the file ? (that's why its 2.9ms instead of 0.8ms?) and for you is it a reasonable speedup ?

**Reply by Tombana** [reply]
> And to clarify we should always look at the line were its written Inference time not the time we see in the end of the file ? (that's why its 2.9ms instead of 0.8ms?)

Correct.
I was confused because the output of this tool is different from what it was in previous tensorflow versions. In previous versions it was always shown at the end of the file but not anymore. I think the 0.8 ms is only the part that was done by the "XNNPack Delegate" (part of the network that is accelerated by the XNNPack library) and 2.9 ms is the full model.

> and for you is it a reasonable speedup ?

I think this is reasonable for this model yes. To get bigger speedups, the layers would need a higher number of channels, and always multiples of 32 channels.

**Reply by christophezeinaty** [reply]
ok well noted, but higher channel means bigger model no ? so if we have memory constraint it wont be the best idea (I guess it should be a comprise between model size and inference time)

**Reply by Tombana** [reply]
That's right.
